### PR TITLE
fix (engine): use execGroups - worker checks wm groups

### PR DESCRIPTION
**Pull request by yesnault**
This will avoid a worker model not shared.infra to run a job not attached to his group

Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

**Comment by sguiheux** [comment]
I agree with @fsamin.  When Hatchery/Worker do a GetQueue, only job that can be executed must be returned. 
So checking the group with requirement is useless

**Comment by yesnault** [comment]
I agree: 
- keep execGroups
- impact getQueue to return only good jobs
- remove checkreq perm in worker
- check perm in take

File: engine/api/workflow/execute_node_run.go [hunk 1]
```diff
@@ -362,6 +370,57 @@ func addJobsToQueue(db gorp.SqlExecutor, stage *sdk.Stage, run *sdk.WorkflowNode
 	return nil
 }
 
+func getJobExecutablesGroups(db gorp.SqlExecutor, run *sdk.WorkflowNodeRun) ([]sdk.Group, error) {
```
Comment (fsamin) [inline engine/api/workflow/execute_node_run.go]: do not return groups, add them to the workflowNodeRun

Comment (fsamin) [inline engine/api/workflow/execute_node_run.go]: Should be nice to have unit tests of this huge sql request

File: engine/worker/run.go [hunk 2]
```diff
@@ -91,9 +91,9 @@ func (w *currentWorker) processActionVariables(a *sdk.Action, parent *sdk.Action
 }
 
 func (w *currentWorker) startAction(ctx context.Context, a *sdk.Action, buildID int64, params *[]sdk.Parameter, stepOrder int, stepName string) sdk.Result {
-	log.Debug("startAction> Begin %p", ctx)
+	log.Debug("startAction> Begin %s", a.ID)
```
Comment (fsamin) [inline engine/worker/run.go]: the logs was here to follow the context propagation; now you can remove it

File: engine/sql/082_wnrj_groups.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+-- +migrate Up
+ALTER TABLE workflow_node_run_job ADD COLUMN exec_groups JSONB;
```
Comment (sguiheux) [inline engine/sql/082_wnrj_groups.sql]: rename fto 083

File: engine/api/workflow/execute_node_run.go (hunk 1 above)
Comment (yesnault) [reply engine/api/workflow/execute_node_run.go]: no groups attr on WorkflowNodeRun, see. https://github.com/ovh/cds/blob/master/sdk/workflow_run.go#L99

File: engine/api/workflow/dao_node_run_job.go [hunk 4]
```diff
@@ -71,7 +71,7 @@ func CountNodeJobRunQueue(db gorp.SqlExecutor, store cache.Store, groupsID []int
 }
 
 // LoadNodeJobRunQueue load all workflow_node_run_job accessible
-func LoadNodeJobRunQueue(db gorp.SqlExecutor, store cache.Store, groupsID []int64, since *time.Time, statuses ...string) ([]sdk.WorkflowNodeJobRun, error) {
+func LoadNodeJobRunQueue(db gorp.SqlExecutor, store cache.Store, isUser bool, groupsID []int64, since *time.Time, statuses ...string) ([]sdk.WorkflowNodeJobRun, error) {
```
Comment (fsamin) [inline engine/api/workflow/dao_node_run_job.go]: is think it's more consistent to pass the user than a boolean

File: engine/api/workflow/dao_node_run_job.go [hunk 5]
```diff
@@ -90,10 +90,29 @@ func LoadNodeJobRunQueue(db gorp.SqlExecutor, store cache.Store, groupsID []int6
 
 	jobs := make([]sdk.WorkflowNodeJobRun, len(sqlJobs))
 	for i := range sqlJobs {
-		getHatcheryInfo(store, &sqlJobs[i])
 		if err := sqlJobs[i].PostGet(db); err != nil {
 			return nil, sdk.WrapError(err, "workflow.LoadNodeJobRun> Unable to load job runs (PostGet)")
 		}
+		var toadd bool
```
Comment (fsamin) [inline engine/api/workflow/dao_node_run_job.go]: this var name is pretty hard to understand this sounds like toad and i don't understand what we have to add

File: engine/api/workflow/execute_node_run.go (hunk 1 above)
Comment (fsamin) [reply engine/api/workflow/execute_node_run.go]: ?

Comment (fsamin) [reply engine/api/workflow/execute_node_run.go]: ok

**Pull request merged**